cc/tfrzz_to_cbt.cc

```diff
@@ -0,0 +1,154 @@
+DEFINE_string(glob_pattern, "", "Input filename glob pattern");
+DEFINE_bool(async, false, "Run in background after incrementing game counter.");
+
+void wait_for_children(std::set<int>* pids, size_t maximum_children) {
+  while (pids->size() > maximum_children) {
+    int status;
+    int child_pid = wait(&status);
+    if (status != 0) {
+      MG_FATAL() << "Child pid " << child_pid << " did not succeed";
+    }
+    auto where = pids->find(child_pid);
+    if (where == pids->end()) {
+      MG_FATAL() << "Child pid " << child_pid << " not found";
+    } else {
+      pids->erase(where);
+    }
+  }
+}
+
+int main(int argc, char* argv[]) {
+  minigo::Init(&argc, &argv);
+
+  std::vector<std::string> bigtable_spec =
+      absl::StrSplit(FLAGS_output_bigtable, ',');
+  bool use_bigtable = bigtable_spec.size() == 3;
+  if (!FLAGS_output_bigtable.empty() && !use_bigtable) {
+    MG_FATAL() << "Bigtable output must be of the form: project,instance,table";
+    return 1;
+  }
+
+  std::deque<std::string> paths(argv + 1, argv + argc);
+  if (!paths.empty()) {
+    std::cout << paths.size() << " files detected on command line."
+              << std::endl;
+  }
+  auto glob_pattern = FLAGS_glob_pattern;
+  if (!glob_pattern.empty()) {
+    std::vector<std::string> glob_expansion;
+    TF_CHECK_OK(tensorflow::Env::Default()->GetMatchingPaths(glob_pattern,
+                                                             &glob_expansion));
+    std::copy(glob_expansion.begin(), glob_expansion.end(),
+              std::back_inserter(paths));
+    std::cout << "Added " << glob_expansion.size() << " files for a total of "
+              << paths.size() << " files to process." << std::endl;
+  }
+
+  std::set<int> pending_children;
+  auto total_games = paths.size();
+  auto const& project = bigtable_spec[0];
+  auto const& instance = bigtable_spec[1];
+  auto const& table = bigtable_spec[2];
+
+  uint64_t final_game_counter = minigo::tf_utils::IncrementGameCounter(
+      project, instance, table, total_games);
+  uint64_t game_counter = final_game_counter - total_games;
+  std::cout << "Initial game counter: " << game_counter << std::endl
+            << "Final game counter will be: " << final_game_counter
+            << std::endl;
+
+  if (FLAGS_async) {
+    // Now let our caller be free to launch the next.
+    if (int handover = fork()) {
+      std::cerr << "PID "
+                << handover
+                << " will continue orchestration."
+                << std::endl;
+      return 0;
+    }
+  }
+
+  int conversion_batch = FLAGS_conversion_batch;
+  auto full_start = absl::Now();
+  while (!paths.empty()) {
+    std::vector<std::string> batch;
+    for (int i = 0; !paths.empty() && i < conversion_batch; ++i) {
+      batch.push_back(paths.front());
+      paths.pop_front();
+    }
+    // Run each batch in a separate process in order to manage the
+    // extraordinary growth in a Bigtable client (may be a bug).
```

Comment:
how big is extraordinary?  why does it make you think it's a bug?

Reply:
Will rebase onto the latest. As to the memory growth, it grows so fast as to cause the process to fail after about 490 games. Because of [this link](https://github.com/GoogleCloudPlatform/google-cloud-go/issues/375), I suspect it may be the same issue, one that also has to be fixed for the C++ client. The gRPC channel setting to turn off tracing is not available in this version (yet). Am reaching out to the developers to confirm. Meanwhile, the separate-process approach also allowed me to greatly increase my conversion rate through multiprocessing, so I had more than one motive.

Reply:
I found the bug: https://github.com/grpc/grpc/issues/15340

and the fix got merged 10 days ago: https://github.com/grpc/grpc/pull/16767

Once this becomes available, I'll need to add the new parameter to the channel arguments. Till then, the fork-per-batch ends up limiting it as well.